mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java

```diff
@@ -14,16 +23,69 @@ public class AnnotationHandlerMapping {
     private final Object[] basePackage;
     private final Map<HandlerKey, HandlerExecution> handlerExecutions;
 
-    public AnnotationHandlerMapping(final Object... basePackage) {
+    public AnnotationHandlerMapping(final Object... basePackage) throws Exception {
         this.basePackage = basePackage;
         this.handlerExecutions = new HashMap<>();
+        initialize();
     }
 
-    public void initialize() {
+    public void initialize() throws Exception {
+        for (Class<?> controllerClass : findControllerClasses()) {
+            initializeHandlerExecutions(controllerClass);
+        }
+
         log.info("Initialized AnnotationHandlerMapping!");
     }
 
     public Object getHandler(final HttpServletRequest request) {
-        return null;
+        String url = request.getRequestURI();
+        RequestMethod requestMethod = RequestMethod.from(request.getMethod());
+
+        HandlerKey handlerKey = new HandlerKey(url, requestMethod);
+        return handlerExecutions.get(handlerKey);
+    }
+
+    private Set<Class<?>> findControllerClasses() {
+        Reflections reflections = new Reflections(basePackage);
+        return reflections.getTypesAnnotatedWith(Controller.class);
+    }
+
+    private void initializeHandlerExecutions(Class<?> controllerClass) throws Exception {
+        List<Method> mapperMethods = findMapperMethods(controllerClass);
+        Object controllerInstance = controllerClass.getDeclaredConstructor().newInstance();
+
+        for (Method mapperMethod : mapperMethods) {
+            addMappings(mapperMethod, controllerInstance);
+        }
+    }
+
+    private List<Method> findMapperMethods(Class<?> clazz) {
+        return Arrays.stream(clazz.getMethods())
+                .filter(method -> method.isAnnotationPresent(RequestMapping.class))
+                .toList();
+    }
+
+    private void addMappings(Method mapperMethod, Object controllerInstance) {
+        RequestMapping annotation = mapperMethod.getAnnotation(RequestMapping.class);
+        RequestMethod[] requestMethods = annotation.method();
+        String uri = annotation.value();
+
+        if (requestMethods.length == 0) {
+            requestMethods = RequestMethod.values();
+        }
+
```

Comment:
> @RequestMapping()에 method 설정이 되어 있지 않으면 모든 HTTP method를 지원해야 한다.

는 요구사항을 구현한 부분이네요. 매우 적절한 구현이지만 구현과 별개로 코드의 의미를 알기 위해 시간이 필요하다는 단점이 있다고 느껴집니다. 아래와 같이 메서드를 분리해서 의미를 분명하게 담아보는건 어떨까요?

```suggestion
private void addMappings(Method mapperMethod, Object controllerInstance) {
    RequestMapping annotation = mapperMethod.getAnnotation(RequestMapping.class);
    String uri = annotation.value();
    
    RequestMethod[] requestMethods = getRequestMethods(annotation);
    // ...
}

// 요청 메서드 목록을 가져오는 메서드
private RequestMethod[] getRequestMethods(RequestMapping annotation) {
    // 메서드가 명시되지 않으면 모든 HTTP 메서드를 기본값으로 사용
    RequestMethod[] requestMethods = annotation.method();
    if (requestMethods.length == 0) {
        return RequestMethod.values()
    }
    return requestMethods;
}
```

Reply:
`@interface`는 `default`라는 옵션으로 디폴트 값을 줄 수 있다는 것을 몰랐었어요.
따라서 현재는 `RequestMapping`으로 옮겼습니다!